spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java

```diff
@@ -224,6 +226,23 @@ public Object next() {
 		}
 	}
 
+	@Override
+	protected void addHeaders(Message<?> message, AbstractIntegrationMessageBuilder<?> builder) {
+		if (message.getPayload() instanceof File) {
+			File file = (File) message.getPayload();
+			builder
+				.setHeaderIfAbsent(FileHeaders.ORIGINAL_FILE, file)
+				.setHeaderIfAbsent(FileHeaders.FILENAME, file.getName());
+
+		}
+		else if (message.getPayload() instanceof String) {
+			File file = new File((String) message.getPayload());
```

Comment:
I fully anticipated that you would raise this issue (I remembered the discussion about the tcp mapper) but, after all the recent work around performance in XD, I feel we have to trust the subclass because this avoids the creation of another map for `putAll()`.

Having the subclass return a map, instead of letting it manipulate the builder directly, also raises some complexity - what if the subclass wants to choose `put()` Vs. `putIfAbsent()` (perhaps even for some, but not all headers), we'd have to add more apis.